include/rmm/device_uvector.hpp

```diff
@@ -167,89 +167,131 @@ class device_uvector {
   }
 
   /**
-   * @brief Performs a synchronous copy of `v` to the specified element in device memory.
+   * @brief Performs an asynchronous copy of `v` to the specified element in device memory.
    *
-   * Because this function synchronizes the stream `s`, it is safe to destroy or modify the object
-   * referenced by `v` after this function has returned.
+   * This specialization for fundamental types is optimized to use `cudaMemsetAsync` when
+   * `host_value` is zero.
+   *
+   * This function does not synchronize stream `s` before returning. Therefore, the object
+   * referenced by `v` should not be destroyed or modified until `stream` has been synchronized.
+   * Otherwise, behavior is undefined.
    *
    * @note This function incurs a host to device memcpy and should be used sparingly.
-   * @note This function synchronizes `stream`.
+   *
+   * @note Calling this function with a literal or other r-value reference for `v` is disallowed
+   *  to prevent the implementation from asynchronously copying from a literal or other implicit
+   *  temporary after it is deleted or goes out of scope.
    *
    * Example:
    * \code{cpp}
    * rmm::device_uvector<int32_t> vec(100, stream);
    *
    * int v{42};
    *
-   * // Copies 42 to element 0 on `stream` and synchronizes the stream
-   * vec.set_element(0, v, stream);
-   *
-   * // It is safe to destroy or modify `v`
+   * // Copies 42 to element 0 on `stream`. Does _not_ synchronize
+   * vec.set_element_async(0, v, stream);
+   * ...
+   * cudaStreamSynchronize(stream);
+   * // Synchronization is required before `v` can be modified
    * v = 13;
    * \endcode
    *
-   *
    * @throws rmm::out_of_range exception if `element_index >= size()`
    *
    * @param element_index Index of the target element
    * @param v The value to copy to the specified element
    * @param s The stream on which to perform the copy
    */
-  void set_element(std::size_t element_index, T const& v, cuda_stream_view s)
+  void set_element_async(std::size_t element_index, value_type const& v, cuda_stream_view s)
   {
     RMM_EXPECTS(
       element_index < size(), rmm::out_of_range, "Attempt to access out of bounds element.");
-    RMM_CUDA_TRY(
-      cudaMemcpyAsync(element_ptr(element_index), &v, sizeof(v), cudaMemcpyDefault, s.value()));
-    s.synchronize_no_throw();
+    if constexpr (std::is_fundamental<value_type>::value) {
```

Comment:
Maybe this can be done using SFINAE?

Reply:
It certainly can, but why would I want to go back to SFINAE? I replaced SFINAE with `if constexpr`, which allows having a single implementation of this function rather than three!